src/main/java/org/embulk/input/jira/client/JiraClient.java

```diff
@@ -0,0 +1,197 @@
+            LOGGER.error(String.format("JIRA return status (%s), reason (%s)", e.getStatusCode(), e.getMessage()));
+            if (e.getStatusCode() == HttpStatus.SC_UNAUTHORIZED) {
+                throw new ConfigException("Could not authorize with your credential.");
+            }
+            else {
+                throw new ConfigException("Could not authorize with your credential due to problem when contacting JIRA API");
+            }
+        }
+    }
+
+    public List<Issue> searchIssues(final PluginTask task, int startAt, int maxResults)
+    {
+        String response = searchJiraAPI(task, startAt, maxResults);
+        JsonParser parser = new JsonParser();
+        JsonObject result = parser.parse(response).getAsJsonObject();
+        return StreamSupport.stream(result.get("issues").getAsJsonArray().spliterator(), false)
+                            .map(jsonElement -> {
+                                JsonObject json = jsonElement.getAsJsonObject();
+                                JsonObject fields = json.get("fields").getAsJsonObject();
+                                Set<Entry<String, JsonElement>> entries = fields.entrySet();
+                                json.remove("fields");
+                                // Merged all properties in fields to the object
+                                for (Entry<String, JsonElement> entry : entries) {
+                                    json.add(entry.getKey(), entry.getValue());
+                                }
+                                return new Issue(json);
+                            })
+                            .collect(Collectors.toList());
+    }
+
+    public int getTotalCount(final PluginTask task)
+    {
+        String response = searchJiraAPI(task, 0, MIN_RESULTS);
+        JsonParser parser = new JsonParser();
+        JsonObject result = parser.parse(response).getAsJsonObject();
+        return result.get("total").getAsInt();
+    }
+
+    private String searchJiraAPI(final PluginTask task, int startAt, int maxResults)
+    {
+        try {
+            return retryExecutor().withRetryLimit(task.getRetryLimit())
+            .withInitialRetryWait(task.getInitialRetryIntervalMillis())
+            .withMaxRetryWait(task.getMaximumRetryIntervalMillis())
+            .runInterruptible(new Retryable<String>()
+            {
+                @Override
+                public String call() throws Exception
+                {
+                    return authorizeAndRequestWithGet(task, JiraUtil.buildSearchUrl(task, startAt, maxResults));
+                }
+
+                @Override
+                public boolean isRetryableException(Exception exception)
+                {
+                    if (exception instanceof JiraException) {
+                        int statusCode = ((JiraException) exception).getStatusCode();
+                        if (statusCode / 100 == 4 && statusCode != HttpStatus.SC_UNAUTHORIZED && statusCode != 429) {
```

Comment:
A little hairy and seems like incorrect logic, unauthorized is retry-able? Put aside the correctness, this could be simplified as:
```
return statusCode / 100 != 4 || statusCode == HttpStatus.SC_UNAUTHORIZED || statusCode == 429;
```

Reply:
Done.
The reason when we retry on 401 is because these could be 401 status when the user try to overloading JIRA server. Which we have to do the fix https://github.com/treasure-data/embulk-input-jira/pull/52 for ruby version.

Reply:
Understood, but leaving a comment to explain that would be very helpful. The retrying status code range is a bit puzzled to me, retrying all except 4xx (surely with the exception of 401 and 429), but I'm don't know enough the JIRA specs to tell.

Reply:
The reason why we retry on 401 is because when we try to have so many requests in once to JIRA APIs (i.e 100 requests). Then the JIRA APIs will return 401 error although the credential is correct. 
Hence we need to retry on such situation.
Otherwise 429 is stand for Too many request. We also need to retry.
Other 4xx error are considered errors.